### ISSUE TITLE
[Can't Install] ERROR: Could not build wheels for spherical-geometry which use PEP 517 and cannot be installed directly

### ISSUE THREAD

**Issue by DaneelAndGiskard** [spacetelescope/spherical_geometry#190]
Can't install, having this log (pip 20.0.2; python 3.7.4):
```
C:\Users\laure>pip install spherical-geometry
Collecting spherical-geometry
  Using cached spherical_geometry-1.2.18.tar.gz (9.2 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Requirement already satisfied: numpy in c:\users\laure\appdata\local\programs\python\python37\lib\site-packages (from spherical-geometry) (1.16.4)
Requirement already satisfied: astropy in c:\users\laure\appdata\local\programs\python\python37\lib\site-packages (from spherical-geometry) (4.0)
Building wheels for collected packages: spherical-geometry
  Building wheel for spherical-geometry (PEP 517) ... error
  ERROR: Command errored out with exit status 1:
   command: 'c:\users\laure\appdata\local\programs\python\python37\python.exe' 'c:\users\laure\appdata\local\programs\python\python37\lib\site-packages\pip\_vendor\pep517\_in_process.py' build_wheel 'C:\Users\laure\AppData\Local\Temp\tmpx2_5maoi'
       cwd: C:\Users\laure\AppData\Local\Temp\pip-install-281ifd5c\spherical-geometry
  Complete output (29 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.7
  creating build\lib.win-amd64-3.7\spherical_geometry
  copying spherical_geometry\graph.py -> build\lib.win-amd64-3.7\spherical_geometry
  copying spherical_geometry\great_circle_arc.py -> build\lib.win-amd64-3.7\spherical_geometry
  copying spherical_geometry\polygon.py -> build\lib.win-amd64-3.7\spherical_geometry
  copying spherical_geometry\vector.py -> build\lib.win-amd64-3.7\spherical_geometry
  copying spherical_geometry\__init__.py -> build\lib.win-amd64-3.7\spherical_geometry
  creating build\lib.win-amd64-3.7\spherical_geometry\tests
  copying spherical_geometry\tests\benchmarks.py -> build\lib.win-amd64-3.7\spherical_geometry\tests
  copying spherical_geometry\tests\setup_package.py -> build\lib.win-amd64-3.7\spherical_geometry\tests
  copying spherical_geometry\tests\test_basic.py -> build\lib.win-amd64-3.7\spherical_geometry\tests
  copying spherical_geometry\tests\test_intersection.py -> build\lib.win-amd64-3.7\spherical_geometry\tests
  copying spherical_geometry\tests\test_shared.py -> build\lib.win-amd64-3.7\spherical_geometry\tests
  copying spherical_geometry\tests\test_union.py -> build\lib.win-amd64-3.7\spherical_geometry\tests
  copying spherical_geometry\tests\test_util.py -> build\lib.win-amd64-3.7\spherical_geometry\tests
  copying spherical_geometry\tests\__init__.py -> build\lib.win-amd64-3.7\spherical_geometry\tests
  creating build\lib.win-amd64-3.7\spherical_geometry\utils
  copying spherical_geometry\utils\__init__.py -> build\lib.win-amd64-3.7\spherical_geometry\utils
  creating build\lib.win-amd64-3.7\spherical_geometry\utils\compat
  copying spherical_geometry\utils\compat\weakref.py -> build\lib.win-amd64-3.7\spherical_geometry\utils\compat
  copying spherical_geometry\utils\compat\weakrefset.py -> build\lib.win-amd64-3.7\spherical_geometry\utils\compat
  copying spherical_geometry\utils\compat\__init__.py -> build\lib.win-amd64-3.7\spherical_geometry\utils\compat
  running build_ext
  building 'spherical_geometry.math_util' extension
  error: Microsoft Visual C++ 14.0 is required. Get it with "Build Tools for Visual Studio": https://visualstudio.microsoft.com/downloads/
  ----------------------------------------
  ERROR: Failed building wheel for spherical-geometry
Failed to build spherical-geometry
ERROR: Could not build wheels for spherical-geometry which use PEP 517 and cannot be installed directly1
```

**Comment by mcara** [comment]
CC: @jhunkeler Could you please take a look at this?

**Comment by jhunkeler** [comment]
```
error: Microsoft Visual C++ 14.0 is required [...]
```

Either you do not have Visual Studio installed, the version you do have is too old, or you're not running `pip` from within a "VS201x Developer Command Prompt".

To install Visual Studio go here: https://visualstudio.microsoft.com/vs/community/

After you've installed the software, open a "Developer Command Prompt" (from the Windows start menu), then run the `pip` installation again from within that prompt.

**Comment by dr-rodriguez** [comment]
I am having a similar issue, but on linux with:
```
python          3.9.2
pip             21.1.1
setuptools      56.0.0
wheel           0.36.2
```
And my error message is slightly different since there is no reference to Visual C++:
```
(caom-py39) -bash-4.1$ pip install spherical-geometry
Collecting spherical-geometry
  Using cached spherical_geometry-1.2.19.tar.gz (9.2 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Requirement already satisfied: numpy in ./miniconda3/envs/caom-py39/lib/python3.9/site-packages (from spherical-geometry) (1.20.1)
Requirement already satisfied: astropy in ./miniconda3/envs/caom-py39/lib/python3.9/site-packages (from spherical-geometry) (4.2.1)
Requirement already satisfied: pyerfa in ./miniconda3/envs/caom-py39/lib/python3.9/site-packages (from astropy->spherical-geometry) (1.7.3)
Building wheels for collected packages: spherical-geometry
  Building wheel for spherical-geometry (PEP 517) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/drodriguez/miniconda3/envs/caom-py39/bin/python /home/drodriguez/miniconda3/envs/caom-py39/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py build_wheel /tmp/tmp6oh27enf
       cwd: /tmp/pip-install-cpinsurm/spherical-geometry_015aa8a74a2f42699801f8c2a3cda187
  Complete output (41 lines):
  /tmp/pip-build-env-gcgsxco4/overlay/lib/python3.9/site-packages/setuptools/dist.py:642: UserWarning: Usage of dash-separated 'upload-dir' will not be supported in future versions. Please use the underscore name 'upload_dir' instead
    warnings.warn(
  /tmp/pip-build-env-gcgsxco4/overlay/lib/python3.9/site-packages/setuptools/dist.py:642: UserWarning: Usage of dash-separated 'show-response' will not be supported in future versions. Please use the underscore name 'show_response' instead
    warnings.warn(
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-3.9
  creating build/lib.linux-x86_64-3.9/spherical_geometry
  copying spherical_geometry/polygon.py -> build/lib.linux-x86_64-3.9/spherical_geometry
  copying spherical_geometry/great_circle_arc.py -> build/lib.linux-x86_64-3.9/spherical_geometry
  copying spherical_geometry/__init__.py -> build/lib.linux-x86_64-3.9/spherical_geometry
  copying spherical_geometry/vector.py -> build/lib.linux-x86_64-3.9/spherical_geometry
  copying spherical_geometry/graph.py -> build/lib.linux-x86_64-3.9/spherical_geometry
  creating build/lib.linux-x86_64-3.9/spherical_geometry/utils
  copying spherical_geometry/utils/__init__.py -> build/lib.linux-x86_64-3.9/spherical_geometry/utils
  creating build/lib.linux-x86_64-3.9/spherical_geometry/tests
  copying spherical_geometry/tests/test_intersection.py -> build/lib.linux-x86_64-3.9/spherical_geometry/tests
  copying spherical_geometry/tests/test_shared.py -> build/lib.linux-x86_64-3.9/spherical_geometry/tests
  copying spherical_geometry/tests/__init__.py -> build/lib.linux-x86_64-3.9/spherical_geometry/tests
  copying spherical_geometry/tests/benchmarks.py -> build/lib.linux-x86_64-3.9/spherical_geometry/tests
  copying spherical_geometry/tests/test_basic.py -> build/lib.linux-x86_64-3.9/spherical_geometry/tests
  copying spherical_geometry/tests/setup_package.py -> build/lib.linux-x86_64-3.9/spherical_geometry/tests
  copying spherical_geometry/tests/test_union.py -> build/lib.linux-x86_64-3.9/spherical_geometry/tests
  copying spherical_geometry/tests/test_util.py -> build/lib.linux-x86_64-3.9/spherical_geometry/tests
  creating build/lib.linux-x86_64-3.9/spherical_geometry/utils/compat
  copying spherical_geometry/utils/compat/__init__.py -> build/lib.linux-x86_64-3.9/spherical_geometry/utils/compat
  copying spherical_geometry/utils/compat/weakrefset.py -> build/lib.linux-x86_64-3.9/spherical_geometry/utils/compat
  copying spherical_geometry/utils/compat/weakref.py -> build/lib.linux-x86_64-3.9/spherical_geometry/utils/compat
  running build_ext
  building 'spherical_geometry.math_util' extension
  creating build/temp.linux-x86_64-3.9
  creating build/temp.linux-x86_64-3.9/libqd
  creating build/temp.linux-x86_64-3.9/libqd/src
  creating build/temp.linux-x86_64-3.9/src
  gcc -pthread -B /home/drodriguez/miniconda3/envs/caom-py39/compiler_compat -Wl,--sysroot=/ -Wno-unused-result -Wsign-compare -DNDEBUG -O2 -Wall -fPIC -O2 -isystem /home/drodriguez/miniconda3/envs/caom-py39/include -fno-semantic-interposition -fPIC -O2 -isystem /home/drodriguez/miniconda3/envs/caom-py39/include -fno-semantic-interposition -fPIC -I/tmp/pip-build-env-gcgsxco4/overlay/lib/python3.9/site-packages/numpy/core/include -Isrc -Ilibqd/include -I/home/drodriguez/miniconda3/envs/caom-py39/include/python3.9 -c libqd/src/bits.cpp -o build/temp.linux-x86_64-3.9/libqd/src/bits.o
  cc1plus: error: unrecognized command line option "-fno-semantic-interposition"
  cc1plus: error: unrecognized command line option "-fno-semantic-interposition"
  cc1plus: warning: unrecognized command line option "-Wno-unused-result"
  error: command '/usr/bin/gcc' failed with exit code 1
  ----------------------------------------
  ERROR: Failed building wheel for spherical-geometry
Failed to build spherical-geometry
ERROR: Could not build wheels for spherical-geometry which use PEP 517 and cannot be installed directly
```

Any thoughts on how best to install in this environment? 
Some of the suggestions I've seen (upgrade pip, wheel, etc) I've already done- I'm running the latest version.

**Comment by nden** [comment]
I wonder if [this ](https://github.com/ContinuumIO/anaconda-issues/issues/12319) is the same error.
Are you using Python installed through conda? If so going to Python==3.8 may solve it.

**Comment by dr-rodriguez** [comment]
Thanks for the link there, it does seem like that could be the underlying issue. I'll explore that a bit more. 
I do have spherical-geometry installed under python 3.8.5, but we are testing to see if we can use python 3.9 as part of JWST development.

**Comment by nden** [comment]
@dr-rodriguez Aah, we (SCSB) have concluded this is a show-stopper for now for the pipeline+Python 3.9+conda.

**Comment by dr-rodriguez** [comment]
OK, thanks good to know. It may be a blocker for us (CAOM) as well, though I think we might only use it in our unit testing.

**Comment by jdavies-st** [comment]
Yeah, this works if you use Anaconda's python 3.9.1 package, but their python 3.9.2+ packages are currently broken if you're trying to build C extensions with an old version (<5.7) of `gcc` (to distill what we found in the issue linked above).

**Comment by jdavies-st** [comment]
Thanks @falkben. That is a functioning workaround, but it's a pretty big overhead to install the full C/Fortran compiling toolchain.  Not sure if it has any knock-on effects on other subsystems.

FWIW, the installation packages this pulls for our RHEL 7 systems (which still use gcc 4.8.5) are:

```
$ conda install -c conda-forge compilers

The following NEW packages will be INSTALLED:

  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu
  binutils           conda-forge/linux-64::binutils-2.35.1-hdd6e379_2
  binutils_impl_lin~ conda-forge/linux-64::binutils_impl_linux-64-2.35.1-h193b22a_2
  binutils_linux-64  conda-forge/linux-64::binutils_linux-64-2.35-h67ddf6f_30
  c-compiler         conda-forge/linux-64::c-compiler-1.1.3-h7f98852_0
  compilers          conda-forge/linux-64::compilers-1.1.3-ha770c72_0
  cxx-compiler       conda-forge/linux-64::cxx-compiler-1.1.3-h4bd325d_0
  fortran-compiler   conda-forge/linux-64::fortran-compiler-1.1.3-h1990efc_0
  gcc_impl_linux-64  conda-forge/linux-64::gcc_impl_linux-64-9.3.0-h70c0ae5_19
  gcc_linux-64       conda-forge/linux-64::gcc_linux-64-9.3.0-hf25ea35_30
  gfortran_impl_lin~ conda-forge/linux-64::gfortran_impl_linux-64-9.3.0-hc4a2995_19
  gfortran_linux-64  conda-forge/linux-64::gfortran_linux-64-9.3.0-hdc58fab_30
  gxx_impl_linux-64  conda-forge/linux-64::gxx_impl_linux-64-9.3.0-hd87eabc_19
  gxx_linux-64       conda-forge/linux-64::gxx_linux-64-9.3.0-h3fbe746_30
  kernel-headers_li~ conda-forge/noarch::kernel-headers_linux-64-2.6.32-h77966d4_13
  libgcc-devel_linu~ conda-forge/linux-64::libgcc-devel_linux-64-9.3.0-h7864c58_19
  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-9.3.0-hff62375_19
  libgfortran5       conda-forge/linux-64::libgfortran5-9.3.0-hff62375_19
  libgomp            conda-forge/linux-64::libgomp-9.3.0-h2828fa1_19
  libstdcxx-devel_l~ conda-forge/linux-64::libstdcxx-devel_linux-64-9.3.0-hb016644_19
  python_abi         conda-forge/linux-64::python_abi-3.9-1_cp39
  sysroot_linux-64   conda-forge/noarch::sysroot_linux-64-2.12-h77966d4_13

The following packages will be UPDATED:

  certifi            pkgs/main::certifi-2020.12.5-py39h06a~ --> conda-forge::certifi-2020.12.5-py39hf3d152e_1
  ld_impl_linux-64   pkgs/main::ld_impl_linux-64-2.33.1-h5~ --> conda-forge::ld_impl_linux-64-2.35.1-hea4e1c9_2
  libgcc-ng           pkgs/main::libgcc-ng-9.1.0-hdf63c60_0 --> conda-forge::libgcc-ng-9.3.0-h2828fa1_19
  libstdcxx-ng       pkgs/main::libstdcxx-ng-9.1.0-hdf63c6~ --> conda-forge::libstdcxx-ng-9.3.0-h6de172a_19
```

But yeah, would be nice if Anaconda folks just fix their buggy Python binaries.

**Comment by falkben** [comment]
Are there plans to publish wheels to PyPI for the latest versions of Python, 3.11 and 3.12, for instance?

**Comment by pllim** [comment]
I do not make plans for this package, will have to ask @mcara or @nden . But I opened #239

**Comment by pllim** [comment]
Please state the architectures you are looking for on that issue. Thanks!

**Comment by pllim** [comment]
FWIW, there are now artifacts (should be wheels, I hope) in https://github.com/spacetelescope/spherical_geometry/actions/workflows/publish-to-pypi.yml?query=branch%3Amaster (click on a green build and scroll down).

The next time this package releases (v1.3?), similar wheels will be pushed out to PyPI. But if you need aarch64, you are out of luck for now due to #247 .

Hope this helps!

**Comment by falkben** [comment]
Thanks for that information. I think it would be too cumbersome to have to refer to the `artifact.zip` URL in our build pipeline, unzip it, then specify the whl file. Additionally, I think the artifact.zip disappears after 90 days on github.

We'd probably be better off just uploading these whl files to an internal pypi repo.

Is 1.3 scheduled for release (rough timeline)?

Any thought on feasibility of re-running the publish to pypi pipeline here for the current release, in order to fill in the whl files for that release on pypi? I know you can upload new whl files for an existing pypi release, but not sure how easy it is to do with the existing github actions workflow file you are using in this project.

**Comment by pllim** [comment]
> Is 1.3 scheduled for release (rough timeline)?

As soon as @BGannon2 give me access to https://pypi.org/project/spherical-geometry/ because I need to generate a token (#249). SPB changed the credentials for those accounts and I do not have access anymore to do it on PyPI.

> feasibility of re-running the publish to pypi pipeline here for the current release

That is quite impossible. I had to do quite a bit of build infrastructure update to get the wheels to work.

**Comment by falkben** [comment]
Okay, thanks for answering all my questions!

**Comment by pllim** [comment]
@falkben , you can try to get the pre-release wheel from https://pypi.org/project/spherical-geometry/1.3.0rc1/#files and see.

`pip install spherical-geometry --pre --upgrade`

**Comment by falkben** [comment]
@pllim I tried it out w/ python 3.11 and ran it through the tesscut CI system. It's looking good. The install used pre-built wheels both on my m1 mac as well as when building the docker images.

I haven't done any thorough analysis to see if there were any regressions between the version we were using (`1.2.23`) and this pre release version, though.

This is huge, thanks! Being able to easily update to 3.11 is great.

**Comment by pllim** [comment]
Great to hear the wheels are working!

**Comment by pllim** [comment]
To close the loop here, @mcara has released 1.3.0 so all those sweet sweet wheels should be on PyPI now even without `--pre`.
